kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/ColumnMapping.java

```diff
@@ -166,14 +235,62 @@ private static DataType convertToPhysicalType(
         return logicalType;
     }
 
-    /** Returns the physical name for a given {@link StructField} */
-    public static String getPhysicalName(StructField field) {
-        if (field.getMetadata().contains(COLUMN_MAPPING_PHYSICAL_NAME_KEY)) {
-            return (String) field
-                .getMetadata()
-                .get(COLUMN_MAPPING_PHYSICAL_NAME_KEY);
-        } else {
-            return field.getName();
+    private static boolean allowMappingModeChange(String oldMode, String newMode) {
+        // only upgrade from none to name mapping is allowed
+        return oldMode.equals(newMode) ||
+                (COLUMN_MAPPING_MODE_NONE.equals(oldMode) &&
+                        COLUMN_MAPPING_MODE_NAME.equals(newMode));
+    }
+
+    /**
+     * For each column/field in a {@link Metadata}'s schema, assign an id using the current
+     * maximum id as the basis and increment from there. Additionally, assign a physical name based
+     * on a random UUID or re-use the old display name if the mapping mode is updated on an
+     * existing table.
+     * @param metadata The new metadata to assign ids and physical names to
+     * @param isNewTable whether this is part of a commit that sets the mapping mode on a new table
+     * @return {@link Metadata} with a new schema where ids and physical names have been assigned
+     */
+    private static Metadata assignColumnIdAndPhysicalName(Metadata metadata, boolean isNewTable) {
+        StructType schema = metadata.getSchema();
+        int maxColumnId = Math.max(
+                Integer.parseInt(metadata.getConfiguration()
```

Comment:
Can we add the MAX_ID property to `TableConfig` and use it to parse? btw Delta-Spark is parsing it as `Long`.

Reply:
do you know why `delta-spark` actually parses this as `Long` instead of an `Int`? Schema Ids should be ints

Reply:
the only reason I haven't added the `MAX_COLUMN_ID` to `TableConfig` because it's an internal table property that should be set by users and `TableConfig` seems to be made for properties that can be explicitly set by users.
I think `TableConfig` would need a concept of whether a property should be settable by a user or not, wdyt?

Reply:
I agree, we need a property in the `TableConfig` to say it is changeable by the use or not. It can be a separate PR.